feat: Add an auto tuner for optimizing upload

sadhansood opened the pull request:
## Description

This commit introduces a new upload `auto-tuner` to dynamically optimize concurrency, providing more consistent and efficient performance than the static `--upload-mode` settings. The tuner implements a simple algorithm: it first aggressively probes for peak network capacity by multiplicatively increasing concurrent uploads, confirms the peak with a verification step to ensure stability, and then locks concurrency at 90% of the optimal level for the remainder of the upload. This strategy allows the client to quickly saturate a connection and maintain high, stable throughput, intelligently operating within the global ceiling set by configurations like `max_data_in_flight`.

## Test plan

Numerous uploads running locally and checking upload performance.

mlegner reviewed (commented):
Really nice addition, @sadhansood.

I have a few questions and some suggestions.

sadhansood commented on crates/walrus-sdk/src/client/auto_tune.rs:
```diff
@@ -0,0 +1,284 @@
+/// The multiplicative factor for increasing permits during the search phase.
+const INCREASE_FACTOR: f64 = 1.5;
+/// The factor to apply to the best-performing permit count to find the stable, locked-in value.
+const LOCK_FACTOR: f64 = 0.9;
```
yes good callout, i will make this change.

sadhansood commented on crates/walrus-sdk/src/client/auto_tune.rs:
```diff
@@ -0,0 +1,284 @@
+pub(crate) struct AutoTuneHandle {
+    inner: Arc<AutoTuneInner>,
+    semaphore: Arc<Semaphore>,
+    initial_permits: usize,
```
You’re right—the semaphore just limits the number of in-flight sliver uploads, but the auto-tuner decides how to adjust permits based on `self.window_bytes`. Because we aggregate raw bytes, larger primary slivers contribute proportionally more to the measured throughput, so the algorithm already weights them implicitly. Lmk if I didn't address your concern fully.

sadhansood commented on crates/walrus-sdk/src/config/communication_config.rs:
```diff
@@ -50,6 +61,12 @@ pub struct ClientCommunicationConfig {
+    /// Optional limit under which the client skips the sliver status check before uploads.
+    ///
+    /// `None` disables the check entirely.
+    pub sliver_status_check_threshold: Option<usize>,
```
Ah, this is outdated now as this was from another change.

sadhansood commented on crates/walrus-sdk/src/config/communication_config.rs:
```diff
@@ -209,10 +233,30 @@ impl CommunicationLimits {
+        let limited_by_connections = allowed_by_bytes >= self.max_concurrent_writes;
+        let concurrency = allowed_by_bytes.min(self.max_concurrent_writes).max(1);
+
+        tracing::info!(
```
done

mlegner commented on crates/walrus-storage-node-client/src/client.rs:
```diff
@@ -763,8 +763,11 @@ impl StorageNodeClient {
-
+            .await
+            .map_err(|error| {
+                tracing::debug!("error storing sliver: {:?}", error);
```
Nit: I think an `inspect_err` would work here as well.

mlegner commented on crates/walrus-sdk/src/client/communication/node.rs:
```diff
@@ -486,21 +510,72 @@ impl NodeWriteCommunication {
+                let completed_at = Instant::now();
+                if let Some(handle) = &self.auto_tune_handle {
+                    handle.record_success(sliver.len(), A::sliver_type(), completed_at);
+                }
```
Question: Is there a reason to have this outside of `self.record_throughput_success`?

sadhansood replied:
no, not really. Fixed

mlegner commented on crates/walrus-sdk/src/config/communication_config.rs:
```diff
@@ -33,6 +35,134 @@ fn default_sliver_status_check_threshold() -> usize {
+    /// Enables the additive-increase / multiplicative-decrease controller.
+    pub enabled: bool,
+    /// Number of completed sliver uploads required to close a measurement window.
+    #[serde(default = "default_auto_tune_window_sample_target")]
```
Nit: AFAIU, these individual `serde(default)` annotations are not needed if the whole struct is annotated with `serde(default)`.

mlegner commented on crates/walrus-sdk/src/client/auto_tune.rs:
```diff
@@ -0,0 +1,373 @@
+
+    pub fn record_success(&self, bytes: usize, sliver_type: SliverType, completed_at: Instant) {
+        let weight = self.weight_for(sliver_type);
+        let mut state = self.inner.state.lock().expect("auto tune mutex poisoned");
```
Nit: Normally we use a formulation describing why the expect should succeed.

```suggestion
        let mut state = self.inner.state.lock().expect("mutex should not be poisoned");
```

sadhansood replied:
done

sadhansood commented on crates/walrus-sdk/src/config/communication_config.rs:
```diff
@@ -33,6 +35,134 @@ fn default_sliver_status_check_threshold() -> usize {
+#[serde(default)]
+pub struct DataInFlightAutoTuneConfig {
+    /// Enables the additive-increase / multiplicative-decrease controller.
+    pub enabled: bool,
```
done

sadhansood commented:
Thank you @mlegner for in depth review! I've addressed all your comments and answered remaining questions. Please feel free to lmk if something still seems off (and i'll take care of it in a follow up).

halfprice commented:
Thanks @sadhansood for this great improvement!

I just have a question: for CLI users, will the turned parameters get reset for every new CLI invocation?

sadhansood commented:
> Thanks @sadhansood for this great improvement!
> 
> I just have a question: for CLI users, will the turned parameters get reset for every new CLI invocation?

@halfprice iiuc, then answer is yes - the parameters are used up during the upload and config values are used for every time `store` is called on the cli.

halfprice commented:
> > Thanks @sadhansood for this great improvement!
> > I just have a question: for CLI users, will the turned parameters get reset for every new CLI invocation?
> 
> @halfprice iiuc, then answer is yes - the parameters are used up during the upload and config values are used for every time `store` is called on the cli.

Does that mean that the auto tuner will be more effective if the user uploads more blob in a single `store` (comparing to multiple `store`, each upload single blob)?

mlegner commented on crates/walrus-sdk/src/client/auto_tune.rs:
```diff
@@ -0,0 +1,373 @@
+}
+
+impl AutoTuneHandle {
+    pub fn new(config: &DataInFlightAutoTuneConfig, allowed_permits: usize) -> Option<Self> {
```
OK, this means that we should tell users to set these relatively high if they enable auto-tuning. Can you add some documentation around this in a follow-up PR?
